aip/0162.md

```diff
@@ -0,0 +1,312 @@
+---
+aip:
+  id: 162
+  state: draft
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+**Note:** We use the word _revision_ to refer to a historical reference for a
+particular resource, and intentionally avoid the term _version_, which refers
+to the version of an API as a whole.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // A new revision is committed whenever the book is changed in any way.
+  string revision_id = 5 [
+    (google.api.field_behavior) = IMMUTABLE,
+    (google.api.field_behavior) = OUTPUT_ONLY];
+
+  // The timestamp that the revision was created.
+  google.protobuf.Timestamp revision_create_time = 6
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The resource **must** contain a `revision_id` field, which **should** be a
+  string and contain a short, automatically-generated hash. A good rule of
+  thumb is the last eight characters of a UUID4.
```

Comment:
What's the recommended approach in a world where tagging is a feature?
Using a UUID4 might lead to collisions or confusion as to whether it's a named tag or a revision_id.